makelib/imagelight.mk

```diff
@@ -0,0 +1,214 @@
+# of multiple working directories or on a CI system with multiple executors. All images
+# tagged with this build registry can safely be untagged/removed at the end of the build.
+ifeq ($(origin BUILD_REGISTRY), undefined)
+BUILD_REGISTRY := build-$(shell echo $(HOSTNAME)-$(ROOT_DIR) | shasum -a 256 | cut -c1-8)
+endif
+
+# In order to reduce built time especially on jenkins, we maintain a cache
+# of already built images. This cache contains images that can be used to help speed
+# future docker build commands using docker's content addressable schemes.
+# All cached images go in in a 'cache/' local registry and we follow an MRU caching
+# policy -- keeping images that have been referenced around and evicting images
+# that have to been referenced in a while (and according to a policy). Note we can
+# not rely on the image's .CreatedAt date since docker only updates then when the
+# image is created and not referenced. Instead we keep a date in the Tag.
+CACHE_REGISTRY := cache
+
+# prune images that are at least this many hours old
+PRUNE_HOURS ?= 48
+
+# prune keeps at least this many images regardless of how old they are
+PRUNE_KEEP ?= 24
+
+# don't actually prune just show what prune would do.
+PRUNE_DRYRUN ?= 0
+
+# the cached image format
+CACHE_DATE_FORMAT := "%Y-%m-%d.%H%M%S"
+CACHE_PRUNE_DATE := $(shell export TZ="UTC+$(PRUNE_HOURS)"; date +"$(CACHE_DATE_FORMAT)")
+CACHE_TAG := $(shell date -u +"$(CACHE_DATE_FORMAT)")
+
+REGISTRY_ORGS ?= docker.io
+IMAGE_ARCHS := $(subst linux_,,$(filter linux_%,$(PLATFORMS)))
+IMAGE_PLATFORMS := $(subst _,/,$(subst $(SPACE),$(COMMA),$(filter linux_%,$(PLATFORMS))))
+IMAGE_PLATFORMS_LIST := $(subst _,/,$(filter linux_%,$(PLATFORMS)))
+IMAGE_PLATFORM := $(subst _,/,$(PLATFORM))
+
+# if set to 1 docker image caching will not be used.
+CACHEBUST ?= 0
+ifeq ($(CACHEBUST),1)
+BUILD_ARGS += --no-cache
+endif
+
+ifeq ($(HOSTOS),Linux)
+SELF_CID := $(shell cat /proc/self/cgroup | grep docker | grep -o -E '[0-9a-f]{64}' | head -n 1)
+endif
+
+# =====================================================================================
+# Image Targets
+
+do.img.clean:
+	@for i in $(CLEAN_IMAGES); do \
+		if [ -n "$$(docker images -q $$i)" ]; then \
+			for c in $$(docker ps -a -q --no-trunc --filter=ancestor=$$i); do \
+				if [ "$$c" != "$(SELF_CID)" ]; then \
+					echo stopping and removing container $${c} referencing image $$i; \
+					docker stop $${c}; \
+					docker rm $${c}; \
+				fi; \
+			done; \
+			echo cleaning image $$i; \
+			docker rmi $$i > /dev/null 2>&1 || true; \
+		fi; \
+	done
+
+# this will clean everything for this build
+img.clean:
+	@$(INFO) cleaning images for $(BUILD_REGISTRY)
+	@$(MAKE) do.img.clean CLEAN_IMAGES="$(shell docker images | grep -E '^$(BUILD_REGISTRY)/' | awk '{print $$1":"$$2}')"
+	@$(OK) cleaning images for $(BUILD_REGISTRY)
+
+img.done:
+	@rm -fr $(IMAGE_TEMP_DIR)
+
+img.cache:
+	@for i in $(CACHE_IMAGES); do \
```

Comment:
Does this mean that we'll need to define the list of `CACHE_IMAGES` in the top level Makefile for a project?

Reply:
I believe if someone was consuming this functionality then that is true. As mentioned above, I am not aware of any of our projects that are doing so.